neurons/_validator/validator_session.py

```diff
@@ -328,6 +359,45 @@ def log_responses(self, responses):
         wandb_logger.safe_log(wandb_log)
         console.print(table)
 
+    def deterministically_shuffle_and_batch_queryable_uids(self, metagraph, filtered_uids):
+        """
+        Pseudorandomly shuffles the list of queryable uids, and splits it in batches to reduce concurrent requests from multiple validators
+
+        """
+        validator_uids = metagraph.total_stake >= 1.024e3
+        validator_index = self.get_validator_index()
+        num_validators = (validator_uids == True).sum().item()
+
+        # batch_duration_in_blocks = (EPOCH_LENGTH / VALIDATOR_SHOULD_QUERY_EACH_BATCH_X_TIMES_PER_EPOCH) // num_validators
+        # batch_duration_in_blocks = (360 / 2) // 17 = 10
+        # thus, if there are 17 validators, they will switch groups every 10 blocks, so every ~2 minutes
+
+        # e.g. current_block = 2993336
+        # 2993336 // batch_duration_in_blocks = 2993336 // 10 = 299333
+        # (299333 + validator_index) % num_validators = (299333 + 4) % 17 = 1
+        # as validator 4, I should then query the sub group 1
+
+        batch_duration_in_blocks = int(EPOCH_LENGTH / VALIDATOR_SHOULD_QUERY_EACH_BATCH_X_TIMES_PER_EPOCH) // num_validators
+        batch_number_since_genesis = self.current_block // batch_duration_in_blocks
+        batch_index_to_query = (batch_number_since_genesis + validator_index) % num_validators
+
+        # We need to pseudorandomly shuffle the filtered_uids so that miner X isn't always in the same batch as miner Y
+        # Which would be unfair and create batches of varying speed/quality
+
+        # To avoid affecting the whole random package, we create an instance of it, which we seed
+        seed = self.current_block // PSEUDO_SHUFFLE_EVERY_X_BLOCK
+        rng = random.Random(seed)
+        shuffled_filtered_uids = list(filtered_uids[:])
+        rng.shuffle(shuffled_filtered_uids)
```

Comment:
Discussed with the team and preference here would be to have validators select random groups based on a random seed as opposed to a deterministic one. Since the SN will be receiving external, organic requests from the API and third parties querying via validators, it is important that miners are able to handle more than one generation at a time and are able to manage dynamic loads.

```suggestion
        shuffled_filtered_uids = list(filtered_uids[:])
        random.shuffle(shuffled_filtered_uids)
```

Reply:
The point of shuffling the miners the same way and to split them in batches is to reduce the randomness penalty when mutliple validators challenge a miner at the same time. If each validator shuffle their list differently, then there is no point in deterministically splitting the resulting list in batches.

To challenge miners fairly, **the concurrent requests should be sent by the same validator**, or else you would still have the problem where a miner is randomly challenged by too many miners and its response time explodes. If the concurrent requests were at least sent to a batch, then all miners would be challenged concurrently so they would be evaluated fairly. **Miners in batch A would all get 10 concurrent challenges** (or maybe half of them would get 9 and the other half 10 for example depending on the total number of requests the validator receives), and that could easily be handled by your API or your valis, but a miner should only belong to one validator at a time. This ensure that the best processor on Earth would stay registered on this subnet, and not get randomly deregistered.

If you randomly shuffle the list, then there would be collisions between batches. Some miners could even end up not receiving any challenge while some get a lot of challenges. There would be no point in merging this PR then as the resulting behavior would be similar to the one that is currently implemented in the main branch, which causes random score decay.

Reply:
To provide an update here;
We're working on modelling the effects of this change to show the overall impact on the SN after implementation. Will provide an overview here as soon as available - we have some changes around this area that are rolling out this week as well, so we're re-evaluating this change in combination with an upcoming validator request distribution update too.

Thanks for your patience and appreciate your contribution and thoughts here, I like the approach and hope we can get this merged into a release in the near future.